### PR TITLE
Improve android-ndk property interface

### PR DESCRIPTION
**Pull request by pcc**
Re-creating #102994 which was closed.

---
PR #105716 added support for NDK r25b, and removed support for r15. Since the switch to r25b would have broken existing r15 users anyway, let's take the opportunity to make the interface more user friendly.

Firstly move the android-ndk property to [build] instead of the targets. This is possible now that the NDK has obsoleted the concept of target-specific toolchains.

Also make the property take the NDK root directory instead of the "toolchains/llvm/prebuilt/<host tag>" subdirectory.

### CODE REVIEW

**Comment by pcc** [comment]
r? @Mark-Simulacrum

Previously approved on the previous PR modulo the changelog update. I've switched to the new method of notifying users of config breakage.

cc @chriswailes @jyn514 @slanterns @JohnCSimon @Dylan-DPC

File: compiler/rustc_codegen_cranelift/scripts/setup_rust_fork.sh [hunk 1]
```diff
@@ -31,7 +31,7 @@ index d95b5b7f17f..00b6f0e3635 100644
 EOF
 
 cat > config.toml <<EOF
-change-id = 115898
+change-id = 116998
```
Comment (onur-ozkan) [inline compiler/rustc_codegen_cranelift/scripts/setup_rust_fork.sh:34]: :point_up:

File: compiler/rustc_codegen_gcc/test.sh [hunk 2]
```diff
@@ -215,7 +215,7 @@ function setup_rustc() {
     rm config.toml || true
 
     cat > config.toml <<EOF
-change-id = 115898
+change-id = 116998
```
Comment (onur-ozkan) [inline compiler/rustc_codegen_gcc/test.sh:218]: :point_up:

File: src/bootstrap/bootstrap_test.py [hunk 3]
```diff
@@ -103,7 +103,7 @@ class GenerateAndParseConfig(unittest.TestCase):
     """Test that we can serialize and deserialize a config.toml file"""
     def test_no_args(self):
         build = serialize_and_parse([])
-        self.assertEqual(build.get_toml("change-id"), '115898')
+        self.assertEqual(build.get_toml("change-id"), '116998')
```
Comment (onur-ozkan) [inline src/bootstrap/bootstrap_test.py:106]: :point_up:

File: src/doc/rustc/src/platform-support/aarch64-unknown-teeos.md [hunk 4]
```diff
@@ -58,7 +58,7 @@ To build a rust toolchain, create a `config.toml` with the following contents:
 
 ```toml
 profile = "compiler"
-change-id = 115898
+change-id = 116998
```
Comment (onur-ozkan) [inline src/doc/rustc/src/platform-support/aarch64-unknown-teeos.md:61]: :point_up:

File: src/doc/rustc/src/platform-support/nto-qnx.md [hunk 5]
```diff
@@ -98,7 +98,7 @@ Example content:
 
 ```toml
 profile = "compiler"
-change-id = 115898
+change-id = 116998
```
Comment (onur-ozkan) [inline src/doc/rustc/src/platform-support/nto-qnx.md:101]: :point_up:

File: src/doc/rustc/src/platform-support/openharmony.md [hunk 6]
```diff
@@ -101,7 +101,7 @@ To build a rust toolchain, create a `config.toml` with the following contents:
 
 ```toml
 profile = "compiler"
-change-id = 115898
+change-id = 116998
```
Comment (onur-ozkan) [inline src/doc/rustc/src/platform-support/openharmony.md:104]: :point_up:

File: src/tools/opt-dist/src/tests.rs [hunk 7]
```diff
@@ -61,7 +61,7 @@ pub fn run_tests(env: &Environment) -> anyhow::Result<()> {
 
     let config_content = format!(
         r#"profile = "user"
-change-id = 115898
+change-id = 116998
```
Comment (onur-ozkan) [inline src/tools/opt-dist/src/tests.rs:64]: :point_up:

**Review by onur-ozkan — COMMENTED** [review]
You don't need to change all of change-ids exists in the rust source.

Please revert the following changes:

**Comment by onur-ozkan** [comment]
Just saw this https://github.com/rust-lang/rust/pull/115898#issuecomment-1773481421.

> You don't need to change all of change-ids exists in the rust source.
>
> Please revert the following changes:

I suppose that changes necessary because tests have started failing?

**Comment by bjorn3** [comment]
> You don't need to change all of change-ids exists in the rust source.

You would get a warning otherwise, right?

**Comment by onur-ozkan** [comment]
> > You don't need to change all of change-ids exists in the rust source.
> 
> 
> 
> You would get a warning otherwise, right?

Yes.

**Comment by pcc** [comment]
Reverted all requested change-id changes.

**Comment by lqd** [comment]
syn opt is currently being noisy

**Comment by onur-ozkan** [comment]
If the reason is unknown, we can revert the PR (#117117) for now and reintroduce it with the performance fix later. These days I am moving my apartment and don't have the time to investigate this.

**Comment by lqd** [comment]
The reason of the regression you mean? I believe it is noise and not a real regression. This benchmark is being noisy on other PRs, and usually has a threshold higher than the current one.

**Comment by onur-ozkan** [comment]
Oh I see. Not related with this PR then?

**Comment by lqd** [comment]
Here's syn opt right now. You can see on the left graph the same noise happening to other PRs.

image

image

It's *unfortunately* not as noisy as it used to be: the rate of its usual ± 4% swings has somehow slowed down recently, it's been less noisy than usual and the threshold is down to 1.7%. (which may have started calming down after the llvm 17.0.3 update)

The threshold should adjust quickly and stop reporting this as a regression if the new behavior is not as erratic.

In the meantime, @rustbot label: +perf-regression-triaged

**Pull request merged**